src/combinatorics/ramsey/infinite.lean

```diff
@@ -0,0 +1,483 @@
+that the second element is strictly greater than the first.
+-/
+def unordered_pairs (S : set ℕ) :=
+  {p : ℕ × ℕ // p.fst < p.snd ∧ S p.fst ∧ S p.snd}
+
+notation `[ℕ]²` := unordered_pairs univ
+notation `[` S `]²` := unordered_pairs S
+
+/--
+Given a number m and a coloring f, the projection of a coloring
+of unordered_pairs to a coloring of numbers is given by fixing
+the first element as m. Note that red is arbitrarily chosen
+as the color assigned to any number below m.
+-/
+def project (f : [ℕ]² → color) (m : ℕ) : ℕ → color :=
+  λ n : ℕ,
+    dite (m < n)
+      (λ h : m < n, f (subtype.mk (m, n) ⟨h, trivial, trivial⟩))
+      (λ _, red)
+
+/--
+The coloring of an unordered pair {a, b}, where a < b,
+agrees with the projection of a applied to b.
+-/
+lemma project_eq (f : [ℕ]² → color) (p : [ℕ]²):
+  f p = project f p.val.fst p.val.snd :=
+begin
+  unfold project,
+  simp [p.property.left]
+end
+
+/--
+Given coloring of natural numbers, a homogeneous set is
+one which is always mapped to the same color.
+-/
+def homogeneous (f : ℕ → color) (H : Inf) :=
+  ∃ c : color, ∀ n, n ∈ H → f n = c
+
+/--
+Given a coloring f, a homogeneous projection is a set of natural numbers
+together with a single natural number.
+The intended use is a set which is homogeneous when projected
+with the given number and coloring.
+-/
+structure homogeneous_proj (f : [ℕ]² → color) extends Inf := (pt : ℕ)
+
+instance (f : [ℕ]² → color) : has_coe (homogeneous_proj f) Inf := ⟨λ c, c.to_Inf⟩
+
+/--
+A homogeneous projection p is refined by another homogeneous projection q if:
+the number given by q is greater than the number given by p,
+the number given by q is an elment of the set given by p,
+the set given by q is contained in the set given by p,
+and the projection using the number given by q is homogeneous
+on the set given by q.
+-/
+def refines {f : [ℕ]² → color} (p q: homogeneous_proj f) :=
+p.pt < q.pt ∧ q.pt ∈ p.s ∧ q.s ⊆ p.s ∧ homogeneous (project f q.pt) q
+
+infix `<<`:50 := refines
+
+local attribute [instance] prop_decidable
+
+/--
+A version of the infinite pigeon hole principle, tailored to our use case.
+-/
+lemma pigeon_hole_principle (f : ℕ → color) (H : Inf):
+  infinite (reds f H) ∨ infinite (blues f H) :=
+begin
+  rw or_iff_not_imp_left,
+  intros redFin,
+  simp [infinite, not_forall, not_exists] at redFin,
+  cases redFin with w Hw,
+  intros x,
+  have x_lt_a : x < x + w + 1, exact lt_succ_sum x w,
+  have w_lt_a : w < x + w + 1,
+    { simp, exact lt_succ_sum w x },
+  let a : ℕ := x + w + 1,
+    cases (H.pf a) with b Hb,
+  have w_lt_b : w < b, exact lt_trans w_lt_a Hb.left,
+  have hrb : f b = red ∨ f b = blue,
+    { cases f b, simp, simp },
+  apply exists.intro b,
+  constructor,
+  { exact lt_trans x_lt_a Hb.left },
+  {
+    cases hrb with hRed hBlue,
+    { exact absurd (and.intro hRed Hb.right) (Hw b w_lt_b) },
+    { exact ⟨hBlue, Hb.right⟩ }
+  }
+end
+
+/--
+This is the key lemma, reducing the Ramsey Theorem for pairs
```

Comment:
As a general rule, docstrings should make sense when read out of context. Because the show up in tooltips when users hover over the name (when it's used in another file) in VScode. Comments that explain the what/why/how of a lemma within the context of a file are very welcome, but should use `--` or `/- only 1 dash -/` format.

Reply:
ok, good to know!